src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5062,7 +5062,19 @@ protected override void VisitForEachExpression(BoundForEachStatement node)
         {
             var expr = node.Expression;
             VisitRvalue(expr);
-            CheckPossibleNullReceiver(expr);
+            CheckPossibleNullReceiver(expr, out bool reportedDiagnostic);
+
+            if (reportedDiagnostic || node.EnumeratorInfoOpt == null)
+            {
+                return;
+            }
+
+            var enumeratorReturnType = node.EnumeratorInfoOpt.GetEnumeratorMethod.ReturnTypeWithAnnotations.ToTypeWithState();
```

Comment:
>node.EnumeratorInfoOpt.GetEnumeratorMethod [](start = 39, length = 42)

We should probably be re-infering the GetEnumerator method. 
For example: `foreach (var item in MethodWithInference(x)) ...`. The `GetEnumerator` method from binding is incorrect (we should account for flow information).

Please add a test and check if we already have an issue for this.

Reply:
Filed [https://github.com/dotnet/roslyn/issues/35151](https://github.com/dotnet/roslyn/issues/35151) to follow up. I tried to address it inline, but my test also exposed another issue with needing to use `ApplyConversion` as well and I decided it would be better to do a follow up PR and get this change in.

---
In reply to: [276860947](https://github.com/dotnet/roslyn/pull/35122#discussion_r276860947) [](ancestors = 276860947)